var/spack/repos/builtin/packages/ginkgo/package.py

```diff
@@ -142,15 +144,23 @@ def cmake_args(self):
 
         spec = self.spec
         from_variant = self.define_from_variant
+        # NOTE: We need to explicitly use this form for CMake options instead of
+        # the `from_variant` method. Please do not change this!
+        #
+        # The reason is that `from_variant` **returns an empty string** when the
+        # option is OFF, but we want to explicitly disable these options, so
+        # that Ginkgo does not try to autodetect which can be on some
+        # compiler/platforms combinations (e.g., oneapi + MPI detection).
         args = [
-            from_variant("GINKGO_BUILD_CUDA", "cuda"),
-            from_variant("GINKGO_BUILD_HIP", "rocm"),
-            from_variant("GINKGO_BUILD_SYCL", "sycl"),
-            from_variant("GINKGO_BUILD_OMP", "openmp"),
-            from_variant("GINKGO_BUILD_MPI", "mpi"),
+            "-DGINKGO_BUILD_CUDA=%s" % ("ON" if "+cuda" in spec else "OFF"),
+            "-DGINKGO_BUILD_HIP=%s" % ("ON" if "+rocm" in spec else "OFF"),
+            "-DGINKGO_BUILD_SYCL=%s" % ("ON" if "+sycl" in spec else "OFF"),
+            "-DGINKGO_BUILD_OMP=%s" % ("ON" if "+openmp" in spec else "OFF"),
+            "-DGINKGO_BUILD_MPI=%s" % ("ON" if "+mpi" in spec else "OFF"),
```

Comment:
I tried this because we should have `-DGINKGO_BUILD_MPI=OFF` (or `ON` for that matter, but from my understanding of the spec it should be off in this case) in the spack build line that you show above. But currently, it does not appear at all. The failure with the `oneapi` build from my understanding is because we don't have the flag to disable MPI given to Ginkgo.

Maybe it's a cache issue instead? But I'm very puzzled as to why the `-DGINKGO_BUILD_MPI` string appears nowhere.

Reply:
Sorry, I think I found the problem and it was somewhere else, in one of the `when` requirements... hopefully now the pipeline passes.

Reply:
@tcojean Which version of `ginkgo` are you building? In this PR this is added:
```python
variant("mpi", default=False, description="Enable MPI support", when="@:1.4.0")
```
If you are using ginkgo at a version where the `mpi` variant does not exist (so, greater than v1.4.0) then you'll get an empty string. You'll never get an empty string when `ginkgo` has `~mpi`.